akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSpec.scala

```diff
@@ -0,0 +1,219 @@
+      }
+    }
+
+    "be cancelled before the underlying Future completes" in {
+      implicit def m = materializer
+
+      assertAllStagesStopped {
+        val promise = Promise[Source[Int, Int]]()
+        val aside = Promise[Int]()
+        val result = Promise[akka.Done]()
+        def futureSource = Source.fromFutureSource(
+          promise.future).map { i ⇒
+          aside.success(i); i // should never occur
+        }.watchTermination[Unit]() {
+          case (_, res) ⇒ result.completeWith(res); ()
+        }
+
+        futureSource.runWith(Sink.cancelled) should ===(NotUsed)
+        result.future.futureValue should ===(akka.Done)
+        aside.future.isCompleted should ===(false)
+      }
+    }
+
+    "fails as the underlying Future is failed" in {
+      implicit def m = materializer
+
+      assertAllStagesStopped {
+        val promise = Promise[Source[Int, Int]]()
+        val result = Promise[akka.Done]()
+        def futureSource = Source.fromFutureSource(promise.future)
+        def sink = Sink.fold[Int, Int](1)(_ * _)
+
+        promise.failure(new Exception("Foo"))
+
+        futureSource.runWith(sink).failed.
+          map(_.getMessage)(m.executionContext).futureValue should ===("Foo")
+      }
+    }
+
+    "applies back-pressure according future completion" in {
+      implicit def m = materializer
+
+      assertAllStagesStopped {
+        val probe = TestSubscriber.probe[Int]()
+        val underlying = Iterator.iterate(1)(_ + 1).take(3)
+        val promise = Promise[Source[Int, NotUsed]]()
+        val first = Promise[Unit]()
+        lazy val futureSource =
+          Source.fromFutureSource(promise.future).map {
+            case 1 ⇒
+              first.success({}); 11
+            case f ⇒ (f * 10) + 1
+          }
+
+        futureSource.runWith(Sink asPublisher true).subscribe(probe)
+        promise.isCompleted should ===(false)
+
+        val sub = probe.expectSubscription()
+
+        promise.success(Source.fromIterator(() ⇒ underlying))
+        first.isCompleted should ===(false)
+
+        sub.request(5)
+        probe.expectNext(11)
+        probe.expectNext(21)
+        probe.expectNext(31)
+        probe.expectComplete()
+
+        first.isCompleted should ===(true)
+      }
+    }
+
+    "be materialized with a failure" in {
+      implicit def m = materializer
+      implicit def ec = m.executionContext
+
+      assertAllStagesStopped {
+        def underlying = Future(Source.single(100L).
+          mapMaterializedValue[String](_ ⇒ sys.error("MatEx")))
+
+        val aside = Promise[Long]()
+        def futureSource: Source[Long, Future[String]] =
+          Source.fromFutureSource(underlying).
+            map { i ⇒ aside.success(i); i }
+
+        def graph = futureSource.toMat(Sink.last) { (m, _) ⇒ m }
+
+        graph.run().failed.map(_.getMessage).futureValue should ===("MatEx")
+        aside.future.futureValue should ===(100L)
+      }
+    }
+  }
+
+  "ActorGraphInterpreter" must {
```

Comment:
As the existing test was based on `fromGraph`, add this one based on the async variant `fromFutureSource` (future graph) to make sure.